rfcs/0035-structured-logging.md

```diff
@@ -0,0 +1,108 @@
+
+# Summary
+
+We will replace most functions of taskcluster-lib-monitor, taskcluster-task-analysis, and our debug
+logs in our services/workers with structured log output. This output can be ingested and
+post-processed by any system we desire.
+
+## Motivation
+
+We have many ways of monitoring/logging/reporting our services and workers. Chief among them are as follows:
+
+* Papertrail for stdout from our services and workers
+  * Mostly comprises `console.log` and `debug` in js and various `fmt.Sprintf` from golang
+* Sentry for various errors/warnings that arise in our services
+* Audit logs (which are sent off to kinesis) for some structured logs that are related to security
+* SignalFX for metrics which we send to them via statsum
+* Historical task analysis
+
+Each of these has positive and negative aspects and generally make sense for the job they are being
+called on to perform. However, this proliferation of formats has major drawbacks.
+
+* We have a lot of code to support all of these systems
+* The knowledge of how to interact with these systems is in a silo with only 1-2 people on the team
+* Correlating events across these disparate sources is difficult and time consuming
+* Using debug logs for anything even vaguely important is dangerous. In attempting to debug other
+  issues, a developer can inadvertently turn off important logs
+* In the context of redeployability, we would either have to require any group maintaining their own
+  installation to either run Sentry/statsum/SignalFX or forego monitoring completely
+* Debug logs need to be carefully crafted to include any information necessary rather than just having
+  this information appended by default
+
+# Details
+
+In the discussion of this rfc, we have defined 6 broad classes of logging:
+
+1. Trying to figure out what happened in some production issue
+2. Sending subsets of the log flow to various consumers (fraud pipeline, audit logging, etc.)
+3. Searching for events matching a particular pattern
+4. Detailed tracing of a test failure or other finely-detailed issue
+5. Monitoring and metrics for various elements of our system
+6. Analysis of usage patterns
+
+We can handle all of these cases with structured logging.
+
+It is important to note that cases 1, 3 and 4 can all include logs from libraries not created
+by the Taskcluster team. They all use node's debug logging directly. We cannot (at least at the
+time of this writing) globally override debug module output to be structured so we will always have
+at least some unstructured output. We can always post-process the output and stick it into a field
+of an object but we will have multi-line logging such as stack traces split across multiple objects
+in that case. This is probably acceptable.
+
+Some of these classes will require some guarantees about the format of our output. We will need to
+have a system in place for ensuring that we do not log invalid lines for audit logs for example.
```

Comment:
Personally I would like that but @djmitche thinks that is too much. I can be convinced either way.

I think I'm happy to do simpler validation so long as we don't end up re-implementing 70% of json-schema ourselves with some sort of complicated implementation.

Reply:
I think that json-schema-verifiable level is probably major overkill.  I think a basic validity check is enough, like `assert(typeof x === 'string')` type checks.

Reply:
I don't have doubts that JSON-Schema is capable of doing this verification, I just think that even a simple json-schema validation running in the log-generating system is probably overkill here.  Managing 20? new schemas which will have to be split between multiple repositories is probably non-trivial.  I suspect each type of log message is likely only to be generated in one location.

There's also the option of doing this sort of checking in a post-processing system.  If we spit out bad messages, the post-processor can be in charge of validating incoming messages and alert on weird patterns which violate our rules.  This would keep all possible schemas in one place and would remove the performance impact on the generating systems.

Reply:
Ignoring implementation details for a moment, I think we want:

1. To be able to document fields of events some place other than in comments due to staleness.
2. To be able to publish docs about events so that people don't have to dig through the code to find what event is what.
3. Some *reasonable* level of assertions that what we've said is in an event is actually there.
4. (optional) Some way to "guarantee" that events are parse-able by downstream tools. For instance this could be something like banning nested dictionaries in all events.
5. (optional) A way to say that every event needs to have specific fields, e.g. `eventId`, `rootUrl`, etc.

Note: I think 4/5 are basically the same thing and also by `(optional)` I think I mean, "brian wants this but is not sure if anyone else does."

Here are other things I'm pondering:

1. When I think of these "specified" events I think of defining them in the service/lib directly and then publishing them. It seems like you both imagine defining them in a centralized place. Any plusses/minuses to each approach you can think of?
2. How specific do we want to get with asserting the contents of fields do we want to get? Do we care that something is a valid hostname rather than just a string? IP addr? How about integer between 400 and 500 for instance. The feeling I get is that you both don't want us to get particularly specific (I think I agree) but let's figure out where we draw the line.
3. One thing to keep in mind is that we've already built the machinery for viewing json-schemas but if we mandate that events need to have a simple form then it shouldn't be too hard to build another viewer.
4. How about any-of/all-of-one-of semantics. Do we see ourselves wanting anything like that?

Reply:
After reading what I just wrote, I think I've convinced myself that we should avoid json-schema and just have our own mini specification format.

However, I only believe this to be true if we outright ban nested objects from events. A top level key can have a value of an object but none of the values in that object can ever be objects themselves. Does that make sense?

Reply:
1. The plus to defining these in a single place is that it's easy to enumerate them without hunting-and-gathering from a variety of sources.  That one place would probably be one of these multi-lingual libraries, so it'd be nice to have something to test that all languages produce the same outputs.  I suspect that defining them in the services/workers where they are used will lead to difficulties "hunting and gathering" all of those definitions to bring them together for docs.

So basically I see this as an implementation detail, with easier implementation in a single place.  If you see another easy implementation, by all means..

2. I don't want to be very specific at all - name and JSON type is enough.

I only barely care about the type, really I want just enough to make sure that if it's supposed to be a string it's not `null`.

I'm not really keen on nested objects *at all* -- one level seems to be one too many.  We could compromise by allowing a value to be of type "object" without any further formal description of what's in that object, but we'll need to be careful not to throw arbitrary-sized data in there (like, no task definitions, please).  Maybe initially we should disallow objects as values, and add them if/when we see the need.

3. I think a viewer for this would be pretty quick-and-dirty (basically just a table).

4. any-of/all-of etc. are sort of "variant records".  Those are pretty hard to deal with rapidly and without causing further errors, and I bet most downstream consumers don't deal well with such data.

Basically, I think the output messages should be as simple as possible, so we can be more confident that they are error-free.

Reply:
I believe that using [MozLog](https://wiki.mozilla.org/Firefox/Services/Logging) will assist in getting you all to this "schema". Specifically, if you look at the [MozLog JSON Schema](https://wiki.mozilla.org/Firefox/Services/Logging#MozLog_JSON_schema), it has some validity checks for the top-level kv-pairs + some light type checking for the values within the `Fields` section (though it does support nested-objects within `Fields`). But of course, what you all are talking about is beyond the current definitions within MozLog.

>I want to avoid the case where someone like AJ is digging through the source code trying to figure out what the `origin` field means and why it's only present in half of the log entries. So basically, a list of top-level fields and their descriptions and type.

> 1. To be able to document fields of events some place other than in comments due to staleness.

> 2. To be able to publish docs about events so that people don't have to dig through the code to find what event is what.

This would be very helpful, especially in (but not limited to) the audit log.